content/en/BestPractices/defend-against-ssrf.md

```diff
@@ -0,0 +1,257 @@
+
+
+
+Let's understand this with a practical code example: 
+
+
+
+```
+
+<?php
+
+
+
+# Check if the 'url' GET variable is set
+
+if (isset($_GET['image_url'])){
+
+$url = $_GET['image_url'];
+
+
+
+# Fetch the image from the user-supplied URL
+
+$file = fopen($url, 'rb');
+
+
+
+# Send proper header for png images
+
+header("Content-Type: image/png");
+
+
+
+# Dump image file
+
+fpassthru($file);
+
+}
+
+
+
+echo 'Please enter image url'
+
+
+
+?>
+
+```
+
+
+
+In the above code, the application accepts user input from the `image_url` parameter, which is passed to the `$url` function without any sanitization. The `header("Content-Type: image/png");` is being checked for `image/png` header type. Lastly, the `$file` passes malicious user-supplied input to the `fpassthru()` function and returns the data to the user. 
+
+
+
+Now, an attacker could attempt to extract metadata by crafting the following malicious payload: 
+
+
+
+- http://127.0.0.1/ssrf.php?image_url=http://169.254.169.254/latest/meta-data/hostname
+
+
+
+This payload `http://169.254.169.254/latest/meta-data/hostname` will return the `hostname` related information. Similarly, an attacker could extract `AWS Secret`, which could be used to gain privileged access to the AWS account and compromise it. 
+
+
+
+In the case of AWS Cloud, attacks on cloud metadata could be mitigated by applying the proper validation on the code level and using a solution like IMDSv2. 
+
+---
+
+Since we have now looked at various exploitation scenarios to gain a thorough understanding of SSRF vulnerabilities, it is crucial to learn how we can mitigate and prevent SSRF attacks. 
+
+
+## Prevention Measures 
+
+
+**1. Disallow unnecessary protocols:** It is recommended to restrict the use of unnecessary protocols such as `file://, gopher://, schema://` and others as they could be used to bypass the restrictions. 
+
+**2. Restrict Verbose Responses from Server:** Always use a non-verbose and generic error message. A verbose error could allow an attacker to perform blind attacks.  
+
+**3. Never Trust User Supplied Input:** Properly validate and sanitize the user-supplied input before passing it to sensitive methods such as `URL Parsers`. While writing the code, user-supplied input should always be considered as 'Unsafe'. 
+
+**4. Implement a strong allowlist:** An allowlist-based validation should be used for the IP addresses and DNS names to which your application requires access. This would prevent an attacker from trying to request unintended resources. 
+
+**5. Implement the principle of least privilege:** The principle of least privilege mentions that only the minimum necessary rights should be assigned to a user that requests access to a resource. The principle also states that those rights should be in effect for the shortest time possible.
+
```

Comment:
Ordinarily, I'd say we want to start a recommendation off with an Imperative word, but I think this is a good exception. The head words start with Implement, but we also need to define the concept of "The Principle of Least Privilege". 

I think "mentions" is a weaker word, and "states" might fit better.